README.md
# S4 - A small and fast stack machine VM

S4 is an simple interpreted system where the source code IS the executable code. There is no compilation needed.

Why S4? There are multiple reasons:

1. Many programming environments use tokens and a large SWITCH statement in a loop to run the user program. In those systems, the tokens (the cases in the SWITCH) are often arbitrarily assigned and are not human-readable, so they have no meaning to the programmer when reading the machine code. Additionally there is a compiler, often something similar to Forth, to work in that environment. In these enviromnents, there is a steep learning curve. The programmer needs to learn the user environment and the hundreds or thousands of user functions in the libraries (or "words" in Forth). In S4, there is only one thing to learn; the source code is the machine code, there is no a compiler, and there aren't thousands of functions/words to learn.

2. I wanted an extremely simple interactive programming environment that could be easily understood and modified. My mantra is that a program should be "intuitively obvious upon casual inspection".

3. I wanted to avoid the need for using a multiple gigabyte tool chain and the edit/compile/run paradigm for developing programs.

4. I wanted a minimal program with a small footprint that could be easily configured and deployed to different development boards via the Ardiuino IDE.

S4 is the result of my work towards those goals.

- The entire system is implemented 4 files: S4.h, S4.cpp, pc-main.cpp, and S4.ino.
- The same code runs on both a Windows PC or development board via the Arduino IDE. 

The reference for S4 is here: https://github.com/CCurl/S4/blob/main/reference.txt

# Building S4

- On the PC, I use Microsoft's Visual Studio (Community edition). 
- For Development boards, I use the Arduino IDE. 
- I do not have an Apple or Linux system, so I haven't tried to compile the program for those environments
- However, being such a simple C program, it should not be difficult to port S4 to those environments.

S4 was inspired by STABLE. See https://w3group.de/stable.html for details on STABLE.
A big thanks to Sandor Schneider for the inspiration for this project.


ccc/history.txt
S@D@.b.
999 1000 :fr
201 :FOW
.
`FOW
W!
S@D@o-
..
D@.
S@.
.
S@D@%-
[$ #c@W@`FW 1+$1-]
`FC
999 500 :fr
999l
...........
:cd
999 500 :fr
201 `FOW W!W@.
S@D@%-1+[$ #c@W@`FW 1+$1-]
$-
.
W@`FC
36#*.
iF
89 55 + .
26#*.
20#*
20 6 *
2*
6 6 *
++.
52 26 *
36 36 *
..
A@A+.
A@A+.
A@A+
....
xZ
A@A+.
..
xZ
:cd
S+@.
S+@.
S+@.
S-
@.
S-\
S-\
S@.
S-@.
S+@.
xZ
:Cd
H@.
H+@.
H-@.
H@.
H+
H-
.
4*c@.
4*1+c@.
7
4*1+c@.
256 38 + .
iR
xZ
xX
999l
:fd
999 1000 :fr
..
:fd
1001c@.
1000c@.
.
F
.
xX
999l
999 1000 :fr
:fd
...
F.
S.bD.
...
F
.
999 1000
:fr
:cd
iA
:cd
xX
999l
:cd
999 444:fr
iR
xX
999l
999 234 :fr
iR
:fd
.



iA
V.
M.
34:nn
:cd
S.bD.
DS -
.
xZ
xX
999l
999 373 :fr
:fd
.
777 1000 :fr
iA
:fd
{fb ISI;DS -[I+ c@:n3 1-]I;}
:fb
. . .
{fb ISI;DS -[I+ c@:nn 1-]I;}
:fb
{fb ISI;DS -[I+ c@:nnb 1-]I;}
:fb
iA
{fb X0X;ISI;DS -[I+ c@:nnb X+16=(0X;r)1-]I;X;}
:fb
{fb X1X;ISI;DS -[I+ c@:nnb X+16=(1X;r)1-]I;X;}
:fb
{fb X1X;ISI;DS -[I+ c@:nnb X+8=(b) X16=(1X;r)1-]I;X;}
:fb
{fb X1X;ISI;DS -[I+ c@:nnb X8=(b) X+16=(1X;r)1-]I;X;}
:fb
#16:md
.
iR
.
DS -.
DS - 8/
.
xZ
1000#`S"hi there"
.
#c@[,1+#c@]
#c@[,1+#c@]\
{zt #c@[,1+#c@]\}
1000:zt



1000 1004
%%-
.
%-
[$ c@:nnb 1+$1-]\
{dm %%>($)%-1+[$ c@:nnb 1+$1-]\}
1000 1004
:dm
1000 1004
%%>($)
%-1+
[$ #c@:nnb 1+$1-]\
{dm %%>($)%-1+[$ #c@:nnb 1+$1-]\}
1000 1006 :dm
1006 1000 :dm
xX
999l
H.
1000 201:fr
201 1000:fr
$\
SD
:db
SD 1-:db
D#10-:db
S
:zt
 {zt #c@[,1+#c@]\}
:zt
.
F.
.
777 200 :fr
.
S
#100+ :db
66,77,
200 50 + 260
:db
200 220 :db
214c@.
200 # 14+ c@ +
.
.
200 14 + # c@ +
c@.
214 c@.
218 c@.
219c@.
220c@.
221 c@.
219 c@ 256* 218 c@ + .
222 14+
c@.
222 c@.

228c@.
DS - .
h36
F.
DS -.
iR
1I;54[I+.b1-]
1I;10[I+.b1-]
I.
IJ;
iR
iA
xZ
Z4/.
.
1024#*4/.
65536 4*
.
Z4*.
123
#10:md
$
10/
#10:md$10/
'0+,'0+,'0+,
h234
.
hdef
#16:md'0+$
10/
10*7+
16/
#16/'0+$16/
#16/'0+$16/
.
7+,
7+,
7+,
hdef
#16/'0+$16/
.
.
hdef
#16:md'0+$16/
#16:md'0+$16/
#16:md'0+$16/
.
7+,7+,7+,
62
#'9>(7+),
{h1 #'9>(7+),}
{hc #'9>(7+),}
{hc '0+#'9>(7+)}
13:hc,
14:hc,
15:hc,
4:hc,
h2d1c
0$
#16:md:hc$16/
#16:md:hc$16/
#16:md:hc$16/
#16:md:hc$16/
.
,,,,
.
100@.
104@.
100@.
iR
25@.
100c@.
2@.
8c@.
9c@.
10c@.
11c@.
h100

.
h1000
.
h10000
.
100c@.
101c@.
102c@.
103c@.
65536 2*.
150`S"jack.bmp"
:cd
200`S"rb"
{zt #c@[\#c@,1+]\}
150:zt
:cd
\
xX
iA
999l
\
xZ
:cd
150`S"jack.bmp"\
150:zt
200`S"rb"\200:zt
150:zt',,b200:zt
`FNzt
xZ
xX
999l
:cd
`S"s4.cpp"
iA
150`S"s4.cpp"
200`S"rt"
150 200 1000 :fr
150:zt
200:zt
150 200 `FO
.
.
150#`S"block.999"200#`S"rt"
`FO
.
18037272`FC
150#`S"block.999"200#`S"rt"1000:fc
:cd
:fr
S:zt
.
150#`S"block.999"200#`S"rt"2000:fc
:fr
.
2000:zt
:fd
:fb
xZ
:cd

s
2+
S2+
m@
.
S
#m@.1+
#m@.1+
#m@.1+
#m@.1+
.
Y
#m@.1+
#m@.1+
#m@.1+
#m@.1+
#m@.1+
#m@.1+
#m@.1+
#m@.1+
256#*.
#m@.1+
#m@.1+
#m@.1+
#m@.1+
#m@.1+
#m@.1+
#m@.1+
#m@.1+
#m@.1+
#m@.1+
#m@.1+
#m@.1+
#m@.1+
\Y
#3+m@256##***
$
#2+m@256#**$
#1+m@256*$
m@
+++
.
{la #3+m@256##***$#2+m@256#**$#1+m@256*$m@+++}
Y:la.
Y4+:la.
Y8+:la.
iA
iM
xZ
Y
#:lm.
4+#:lm.
4+#:lm.
4+#:lm.
4+#:lm.
4+#:lm.
4+#:lm.
4+#:lm.
4+#:lm.
4+#:lm.
iA
.
Y
#:lm.4+
#:lm.4+
#:lm.4+
#:lm.4+
#:lm.4+
#:lm.4+
#:lm.4+
#:lm.4+
#:lm.4+
#:lm.4+
iR
472 344 -  2/ 4/
2*
.
472 344 -  4/
\
.
256##**
h10000
h1000000
h1000000
\\\\\\
xX
999l
Y
:cd
:lm
.
Y
#m@
h100000
h10*
xX
999l
Y
:lm
.
Y
#m@.1+
#m@.1+
#m@.1+
#m@.1+
Y
#2+m@h10000*
\\\\\
xX
999L
.l
Y
#:ml.4+
#:ml.4+
.
Y
#:lm.4+
#:lm.4+
#:lm.4+
#:lm.4+
#:lm.4+
:cd
1
256##***
\1
h1000000*
.
xZ
:cd
H.
iA
:cd
H.
150`S"Jack.bmp"200`S"rb"
150:zt200:zt
150:zt',,200:zt
150 200 1000 :fr
F.
iR
:cd
:fb
SD.b.
:Cs
:cd
Y=
.
Y.
iR
:cs:sir:cd
:cs:sir:cd
3 20
%-1+[$ #d@:nnb 1+$1-]
0 25%-1+[$ #d@:nnb 1+$1-]
105,65,
:csiRiFr:cd
:csiRriFr:cd
:csiRriFrr:cd
\\
0 25[$ ##65+,@.b.1+$1-]
0 25
[$1+$1-]\
0 25[$ #.b#@.b  1+$1-]\
0 25[$ #65+.b#@.r 1+$1-]\
0 25[$ #65+,b#@.r 1+$1-]\
0 26[$ #65+,b#@.r 1+$1-]\
0 26[$ #65+,b#@.bbbb 1+ #5:md0=(r) $1-]\
0 26[$ #65+,": "#@.bbbb 1+ #5:md0=(r) $1-]\
0 25[$ #65+,": "#@.r 1+$1-]\
iR
0 26[$ #65+,": "#@.r 1+$1-]\
:cd
0 26[$ #65+,": "#@.r 1+$1-]\
123
..
23
100
#
.
$%
%
..
$
%%
<(b)
[%%<(b)10/]\.
23 100[%%<(b)10/].
3 100[%%<(b)10/].
3 100[%%<(0.)10/].
3 1000[%%<(0.)10/].
{rj [%%<(b)10/].}
23 1000:rj
23 100000:rj
{lj %.[%%<(b)10/]\}
23 100000:lj"*"
23 1000000000:lj"*"
23 100000000:lj"*"
0 26[$ #65+,": "#@10000000:lj 1+$1-]\
0 26[$ #65+,": "#@10000000:lj 1+ #5:md0=(r) $1-]\
0 26[$ #65+,": "#@1000000000:lj 1+ #5:md0=(r) $1-]\
0 26[$ #65+,": "#@10000000000:lj 1+ #5:md0=(r) $1-]\
0 26[$ #65+,": "#@100000000000:lj 1+ #5:md0=(r) $1-]\
0 26[$ #65+,": "#@1000000000:lj 1+ #5:md0=(r) $1-]\
:cd
F
1000000000
%%<
()10/
>
{lj %.b[%%<(b)10/]\}
0 26[$ #65+,": "#@1000000000:lj 1+ #5:md0=(r) $1-]\
0 26[$ #65+,": "#@1000000000:lj 1+ #5:md0=(r) $1-]\
0 26[$ #65+,": "#@1000000000:lj 1+ #5:md0=(r) $1-]\b
0 26[$ #65+,": "#@1000000000:ljb 1+ #5:md0=(r) $1-]\b
0 26[$ #65+,": "#@1000000000:rj 1+ #5:md0=(r) $1-]\
0 26[$ #65+,": "#@1000000000:rjbb 1+ #5:md0=(r) $1-]\
:cd
{rj 10/[%%<(b)10/].}
0 26[$ #65+,": "#@1000000000:rjbb 1+ #5:md0=(r) $1-]\
{rj 10/[%%<(0.)10/].}
0 26[$ #65+,": "#@1000000000:rjbb 1+ #5:md0=(r) $1-]\
{rj [%%<(b)10/].}
0 26[$ #65+,": "#@1000000000:rjbb 1+ #5:md0=(r) $1-]\
{rj [%%<(0.)10/].}
0 26[$ #65+,": "#@1000000000:rjbb 1+ #5:md0=(r) $1-]\
{rj [%%<(b)10/#10<].}
0 26[$ #65+,": "#@1000000000:rjbb 1+ #5:md0=(r) $1-]\
.
{rj [%%<(b)10/#10>].}
\\
0 26[$ #65+,": "#@1000000000:rjbb 1+ #5:md0=(r) $1-]\
:cd
200`S"jack.bmp"250`S"rb"
200:zt250:zt
200 250 1000 :fr
200`S"block.999"250`S"rt"

s4.cpp
// S4 - a stack VM, inspired by Sandor Schneider's STABLE - https://w3group.de/stable.html

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdarg.h>
#include "s4.h"

#define STK_SZ 31

struct {
    long dsp, rsp;
    byte code[CODE_SZ];
    long mem[MEM_SZ];
    int func[NUM_FUNCS];
    long dstack[STK_SZ + 1];
    long rstack[STK_SZ + 1];
} sys;

byte isBye = 0, isError = 0;
char buf[100];
FILE* input_fp = NULL;
byte* bMem;

#define CODE       sys.code
#define MEM        sys.mem
#define FUNC       sys.func


#define T        sys.dstack[sys.dsp]
#define N        sys.dstack[sys.dsp-1]
#define R        sys.rstack[sys.rsp]
#define HERE     MEM[7]

void push(long v) { if (sys.dsp < STK_SZ) { sys.dstack[++sys.dsp] = v; } }
long pop() { return (sys.dsp > 0) ? sys.dstack[sys.dsp--] : 0; }

void rpush(long v) { if (sys.rsp < STK_SZ) { sys.rstack[++sys.rsp] = v; } }
long rpop() { return (sys.rsp > 0) ? sys.rstack[sys.rsp--] : -1; }

void vmReset() {
    sys.dsp = sys.rsp = 0;
    for (int i = 0; i < CODE_SZ; i++) { CODE[i] = 0; }
    for (int i = 0; i < MEM_SZ; i++) { MEM[i] = 0; }
    for (int i = 0; i < NUM_FUNCS; i++) { FUNC[i] = 0; }
    MEM['C' - 'A'] = CODE_SZ;
    MEM['D' - 'A'] = (long)&sys.code[0];
    MEM['F' - 'A'] = (long)&sys.func[0];
    MEM['M' - 'A'] = (long)&sys.mem[0];
    MEM['N' - 'A'] = NUM_FUNCS;
    MEM['R' - 'A'] = (long)&sys.rstack[0];
    MEM['S' - 'A'] = (long)&sys.dstack[0];
    MEM['Y' - 'A'] = (long)&sys;
    MEM['Z' - 'A'] = MEM_SZ;
}

void vmInit(int code_sz, int mem_sz, int num_funcs) {
    bMem = (byte*)&sys.mem[0];
    vmReset();
}

void printStringF(const char* fmt, ...) {
    va_list args;
    va_start(args, fmt);
    vsnprintf(buf, sizeof(buf), fmt, args);
    va_end(args);
    printString(buf);
}

int hexNum(char x) {
    if (('0' <= x) && (x <= '9')) { return x - '0'; }
    if (('A' <= x) && (x <= 'F')) { return x - 'A' + 10; }
    if (('a' <= x) && (x <= 'f')) { return x - 'a' + 10; }
    return -1;
}

int funcNum(char x, int alpha) {
    if (('A' <= x) && (x <= 'Z')) { return x - 'A'; }
    if (('a' <= x) && (x <= 'z')) { return x - 'a' + 26; }
    if (('0' <= x) && (x <= '9') && (alpha)) { return x - '0' + 52; }
    isError = 1;
    return -1;
}

int GetFunctionNum(int pc, int msg) {
    int f1 = funcNum(CODE[pc], 0);
    int f2 = funcNum(CODE[pc + 1], 1);
    if ((f1 < 0) || (f2 < 0)) {
        if (msg) { printStringF("-%c%c:FN Bad-", CODE[pc], CODE[pc + 1]); }
        isError = 1;
        return -1;
    }
    int fn = (f1 * 62) + f2;
    if ((fn < 0) || (NUM_FUNCS <= fn)) {
        if (msg) { printStringF("-%c%c:FN OOB-", CODE[pc], CODE[pc + 1]); }
        isError = 1;
        return -1;
    }
    return fn;
}

int doDefineFunction(int pc) {
    if (pc < HERE) { return pc; }
    int fn = GetFunctionNum(pc, 1);
    if (fn < 0) { return pc + 2; }
    CODE[HERE++] = '{';
    CODE[HERE++] = CODE[pc];
    CODE[HERE++] = CODE[pc + 1];
    pc += 2;
    FUNC[fn] = HERE;
    while ((pc < CODE_SZ) && CODE[pc]) {
        CODE[HERE++] = CODE[pc++];
        if (CODE[HERE - 1] == '}') { return pc; }
    }
    isError = 1;
    printString("-overflow-");
    return pc;
}

void dumpCode() {
    printStringF("\r\nCODE: size: %d bytes, HERE=%d", CODE_SZ, HERE);
    if (HERE == 0) { printString("\r\n(no code defined)"); return; }
    int ti = 0, x = HERE, npl = 20;
    char txt[32];
    for (int i = 0; i < HERE; i++) {
        if ((i % npl) == 0) {
            if (ti) { txt[ti] = 0;  printStringF(" ; %s", txt); ti = 0; }
            printStringF("\n\r%05d: ", i);
        }
        txt[ti++] = (CODE[i] < 32) ? '.' : CODE[i];
        printStringF(" %3d", CODE[i]);
    }
    while (x % npl) {
        printString("    ");
        x++;
    }
    if (ti) { txt[ti] = 0;  printStringF(" ; %s", txt); }
}

void dumpFuncs() {
    printStringF("\r\nFUNCTIONS: (%d available)", NUM_FUNCS);
    int n = 0;
    for (int i = 0; i < NUM_FUNCS; i++) {
        if (FUNC[i]) {
            byte f1, f2, x = (i / 62);
            f1 = 'A' + x;
            if (26 <= x) { f1 = 'a' + (x-26); }
            x = (i % 62);
            f2 = 'A' + x;
            if (26 <= x) { f2 = 'a' + (x-26); }
            if (52 <= x) { f2 = '0' + (x-52); }
            if (((n++) % 6) == 0) { printString("\r\n"); }
            printStringF("%c%c:%4d (%d)    ", f1, f2, FUNC[i], i);
        }
    }
}

void dumpStack(int hdr) {
    if (hdr) { printStringF("\r\nSTACK: size: %d ", STK_SZ); }
    printString("(");
    for (int i = 1; i <= sys.dsp; i++) { printStringF("%s%ld", (i > 1 ? " " : ""), sys.dstack[i]); }
    printString(")");
}

void dumpMemory(int isRegs) {
    int c = 0, n = isRegs ? 26 : MEM_SZ;
    if (isRegs) { printStringF("\r\nREGISTERS: "); }
    else { printStringF("\r\nMEMORY: size: %d bytes (%d longs)", MEM_SZ * 4, MEM_SZ); }
    for (int i = 0; i < n; i++) {
        long x = MEM[i];
        if ((!x) && (!isRegs)) { continue; }
        if (((c++) % 5) == 0) { printString("\r\n"); }
        if (isRegs) { printStringF("%c: %-10ld  ", i + 'A', x); }
        else { printStringF("[%05d]: %-10ld  ", i, x); }
    }
    if (c == 0) { printString("\r\n(all memory empty)"); }
}

void dumpAll() {
    dumpStack(1);   printString("\r\n");
    dumpMemory(1);  printString("\r\n");
    dumpCode();     printString("\r\n");
    dumpFuncs();    printString("\r\n");
}

int doFile(int pc) {
    int ir = CODE[pc++];
    switch (ir) {
#ifdef __PC__
    case 'C':
        if (T) { fclose((FILE*)T); }
        pop();
        break;
    case 'O': {
        char* md = (char*)bMem + pop();
        char* fn = (char*)bMem + T;
        T = 0;
        fopen_s((FILE**)&T, fn, md);
    }
            break;
    case 'R': if (T) {
        long n = fread_s(buf, 2, 1, 1, (FILE*)T);
        T = ((n) ? buf[0] : 0);
        push(n);
    }
            break;
    case 'W': if (T) {
        FILE* fh = (FILE*)pop();
        buf[1] = 0;
        buf[0] = (byte)pop();
        fwrite(buf, 1, 1, fh);
    }
            break;
#endif
    case 'N':
        push(0);
        ir = GetFunctionNum(pc, 1);
        if (0 <= ir) { T = FUNC[ir]; }
        pc += 2;
        break;
    }
    return pc;
}

int doPin(int pc) {
    int ir = CODE[pc++];
    long pin = pop(), val = 0;
    switch (ir) {
    case 'I': pinMode(pin, INPUT); break;
    case 'U': pinMode(pin, INPUT_PULLUP); break;
    case 'O': pinMode(pin, OUTPUT); break;
    case 'R': ir = CODE[pc++];
        if (ir == 'D') { push(digitalRead(pin)); }
        if (ir == 'A') { push(analogRead(pin)); }
        break;
    case 'W': ir = CODE[pc++]; val = pop();
        if (ir == 'D') { digitalWrite(pin, val); }
        if (ir == 'A') { analogWrite(pin, val); }
        break;
    }
    return pc;
}

int doExt(int pc) {
    byte ir = CODE[pc++];
    switch (ir) {
    case 'F': pc = doFile(pc);          break;
    case 'X': vmReset();                break;
    case 'P': pc = doPin(pc);           break;
    case 'T': isBye = 1;                break;
    }
    return pc;
}


int step(int pc) {
    byte ir = CODE[pc++];
    long t1, t2;
    switch (ir) {
    case 0: return -1;                                  // 0
    case ' ': while (CODE[pc] == ' ') { pc++; }         // 32
        break;
    case '!': t2 = pop(); t1 = pop();                   // 33
        if ((0 <= t2) && (t2 < MEM_SZ)) { MEM[t2] = t1; }
        break;
    case '"': buf[1] = 0;                          // 34
        while ((pc < CODE_SZ) && (CODE[pc] != '"')) {
            buf[0] = CODE[pc++];
            printString(buf);
        }
        ++pc; break;
    case '#': push(T);               break;             // 35 (DUP)
    case '$': t1 = N; N = T; T = t1; break;             // 36 (SWAP)
    case '%': push(N);               break;             // 37 (OVER)
    case '\\': pop();                break;             // 92 (DROP)
    case '&': t1 = pop(); T &= t1;   break;             // 38
    case '\'': push(CODE[pc++]);     break;             // 39
    case '(': if (pop() == 0) {                         // 40
        while ((pc < CODE_SZ) && (CODE[pc] != ')')) { ++pc; }
        ++pc;
    }
            break;
    case ')': /*maybe ELSE?*/        break;             // 41
    case '*': t1 = pop(); T *= t1;   break;             // 42
    case '+': t1 = pop(); T += t1;   break;             // 43
    case '-': t1 = pop(); T -= t1;   break;             // 45
    case '/': t1 = pop(); if (t1) { T /= t1; }  break;  // 47
    case ',': printStringF("%c", (char)pop());  break;  // 44
    case '.': printStringF("%ld", pop());       break;  // 46
    case '0': case '1': case '2': case '3': case '4':   // 48-57
    case '5': case '6': case '7': case '8': case '9':
        push(ir - '0');
        t1 = CODE[pc] - '0';
        while ((0 <= t1) && (t1 <= 9)) {
            T = (T * 10) + t1;
            t1 = CODE[++pc] - '0';
        }
        break;
    case ':': t1 = GetFunctionNum(pc, 1);                // 58
        pc += 2;
        if ((0 <= t1) && (FUNC[t1])) { 
            rpush(pc);
            pc = FUNC[t1];
        } 
        break;
    case ';': pc = rpop();                       break;  // 59
    case '<': t1 = pop(); T = T < t1  ? -1 : 0;  break;  // 60
    case '=': t1 = pop(); T = T == t1 ? -1 : 0;  break;  // 61
    case '>': t1 = pop(); T = T > t1  ? -1 : 0;  break;  // 62
    // case '?': push(_getch());                    break;  // 63
    case '@': if ((0 <= T) && (T < MEM_SZ)) { T = MEM[T]; }
            break;
    case 'A': case 'B': case 'C': case 'D': case 'E': case 'F':
    case 'G': case 'H': case 'I': case 'J': case 'K': case 'L':
    case 'M': case 'N': case 'O': case 'P': case 'Q': case 'R':
    case 'S': case 'T': case 'U': case 'V': case 'W': case 'X':
    case 'Y': case 'Z': ir -= 'A';
        push(MEM[ir]); t1 = CODE[pc];
        if (t1 == '+') { ++pc; ++MEM[ir]; }
        if (t1 == '-') { ++pc; --MEM[ir]; }
        if (t1 == ';') { pop(); ++pc; MEM[ir] = pop(); }
        break;
    case '[': rpush(pc);                                // 91
        if (T == 0) {
            while ((pc < CODE_SZ) && (CODE[pc] != ']')) { pc++; }
        }
        break;
    case ']': if (T) { pc = R; }                        // 93
            else { pop();  rpop(); }
            break;
    case '^': t1 = pop(); T ^= t1;      break;          // 94
    case '_':                                           // 95
        while (CODE[pc] && (CODE[pc] != '_')) { bMem[T++] = CODE[pc++]; }
        ++pc; bMem[T++] = 0;
        break;
    case '`': /* FREE*/                 break;          // 96
    case 'b': printString(" ");         break;
    case 'c': ir = CODE[pc++];
        t1 = pop();
        if ((0 <= t1) && (t1 < (MEM_SZ * 4))) {
            if (ir == '@') { push(bMem[t1]); }
            if (ir == '!') { bMem[t1] = (byte)pop(); }
        }
        break;
    case 'd': ir = CODE[pc++];
        t1 = pop();
        if ((0 <= t1) && (t1 < CODE_SZ)) {
            if (ir == '@') { push(CODE[t1]); }
            if (ir == '!') { CODE[t1] = (byte)pop(); }
        }
        break;
    case 'e': rpush(pc); pc = pop();    break;
    case 'h': push(0);
        t1 = hexNum(CODE[pc]);
        while (0 <= t1) {
            T = (T * 0x10) + t1;
            t1 = hexNum(CODE[++pc]);
        }
        break;
    case 'i': t1 = CODE[pc++];
        if (t1 == 'A') { dumpAll(); }
        if (t1 == 'C') { dumpCode(); }
        if (t1 == 'F') { dumpFuncs(); }
        if (t1 == 'M') { dumpMemory(0); }
        if (t1 == 'R') { dumpMemory(1); }
        if (t1 == 'S') { dumpStack(0); }
        break;
    case 'j': pc = pop();               break;
    case 'l': t1 = pop();               // LOAD
#ifdef __PC__
        if (input_fp) { fclose(input_fp); }
        sprintf_s(buf, sizeof(buf), "block.%03ld", t1);
        fopen_s(&input_fp, buf, "rt");
#else
        printString("-l:pc only-");
#endif
        break;
    case 'm': ir = CODE[pc++]; {
        byte* bp = (byte*)T;
        if (ir == '@') { T = *bp; }
        if (ir == '!') { t1 = pop(); t2 = pop(); bp[t1] = (byte)t2; }
    } break;
    case 'n': printString("\r\n");      break;
    case 'r': ir = CODE[pc++]; {
        if (ir == '<') { rpush(pop()); }
        if (ir == '@') { push(R); }
        if (ir == '>') { push(rpop()); }
    } break;
    case 't': push(millis());            break;
    case 'w': delay(pop());              break;
    case 'x': pc = doExt(pc);            break;
    case '{': pc = doDefineFunction(pc); break;    // 123
    case '|': t1 = pop(); T |= t1;       break;    // 124
    case '}': pc = rpop();               break;    // 125
    case '~': T = ~T;                    break;    // 126
    }
    return pc;
}

int run(int pc) {
    isError = 0;
    while (sys.rsp >= 0) {
        if ((pc < 0) || (CODE_SZ <= pc)) { return pc; }
        pc = step(pc);
        if (isError) { break; }
    }
    return pc;
}

void setCodeByte(int addr, char ch) {
    if ((0 <= addr) && (addr < CODE_SZ)) { CODE[addr] = ch; }
}

long registerVal(int reg) {
    if ((0 <= 'A') && (reg <= 'Z')) { return MEM[reg - 'A']; }
    if ((0 <= 'a') && (reg <= 'z')) { return MEM[reg - 'a']; }
    return 0;
}

int functionAddress(const char* fname) {
    int pc = registerVal('H') + 2;
    CODE[pc] = fname[0];
    CODE[pc+1] = fname[1];
    int fn = GetFunctionNum(pc, 0);
    return (fn < 0) ? fn : FUNC[fn];
}
